### PR TITLE
Add an LLBC backend

### PR DESCRIPTION
**Pull request by zhassan-aws**
This PR adds a new codegen backend that generates low-level borrow calculus (LLBC), which is the format defined by [Charon](https://github.com/AeneasVerif/charon) and [Aeneas](https://github.com/AeneasVerif/aeneas).

The backend can be invoked using `-Zaeneas`, and will generate a `.llbc` file. This file can then be passed to Aeneas to generate Lean for example. Currently, Aeneas needs to be manually run on the generated LLBC.

The PR translates stable MIR to unstructured LLBC (ULLBC) and then uses the Charon library to apply transformations that reconstruct the control flow to regain some of the high-level structure (loops, if statements, etc.).

In this PR, very few MIR constructs are handled, but the PR is already pretty big. More support to come in subsequent PRs.

Call-outs:

- Charon is currently not released on crates.io, so the library is added as a dependency through GitHub with a particular commit.
- The Kani driver is currenty heavily tailored towards CBMC, so currently it's expecting a goto file, and thus errors out when the Aeneas backend is invoked.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by tautschnig — APPROVED** [review]
Approving with the following understanding:
1. Some cleanup is needed as indicated by the comments.
2. The codegen for Aeneas does need further work, but this doesn't impact any existing user who continues to use the cprover back-end.

File: kani-compiler/src/args.rs
Comment (celinval) [inline kani-compiler/src/args.rs:18]: Why none? Shouldn't we make the argument optional instead?

File: kani-compiler/src/args.rs
Comment (celinval) [inline kani-compiler/src/args.rs:11]: This option should only be available if the compiler feature was enabled, right?

**Review by celinval — COMMENTED** [review]
I haven't reviewed the backend code yet, but I'm curious about how this affects our released software. Are we shipping the new backend?

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (Nadrieril) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:227]: That's a comment copied from Charon, "Rk" is short for @R1kM I believe

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (Nadrieril) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:257]: This is copied over from Charon, I think the answer was "whenever we understand why we'd need the name of an opaque type". We could probably make this an error? I haven't tested if this can ever occur with the current code.

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (R1kM) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:227]: I believe it stands for "Remark" here

File: kani-compiler/src/args.rs
Comment (zhassan-aws) [reply kani-compiler/src/args.rs:11]: Good point.

File: kani-compiler/src/args.rs
Comment (zhassan-aws) [reply kani-compiler/src/args.rs:18]: I think I was trying to simplify the backend checking logic under different features. I'll clean this up.

**Review by zhassan-aws — COMMENTED** [review]
> I haven't reviewed the backend code yet, but I'm curious about how this affects our released software. Are we shipping the new backend?

That was the plan (it's under an unstable feature). If we prefer to not include it in the release at this point, we can disable the feature by default and have a separate set of regressions for this feature.

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (Nadrieril) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:227]: Hah that makes so much more sense x)

File: kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs
Comment (zhassan-aws) [reply kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs]: Removed this one

Comment (zhassan-aws) [reply kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs]: I made the printing of the final LLBC dependent on a new option (`--print-llbc`) that is used in the current expected tests. Now, by default, nothing will get printed except the name of the LLBC file.

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (zhassan-aws) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:227]: Fun conversation :)

I was hesitant to remove the comments because we may need to import any future upstream changes to this function, so I wanted to minimize the diff. @tautschnig Let me know if you think the comments create confusion, and I'll remove them.

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (zhassan-aws) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:257]: Same comment here @tautschnig

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (zhassan-aws) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs]: Expanded comment.

File: kani-compiler/Cargo.toml
Comment (celinval) [reply kani-compiler/Cargo.toml:13]: I'm curious. Why not use a git submodule?

File: kani-compiler/src/args.rs
Comment (celinval) [inline kani-compiler/src/args.rs:90]: Why would we print to the stdout?

File: kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs
Comment (celinval) [inline kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs:217]: I'm curious. Do we want to keep this model for Aeneas backend too? Not something that we need to change for this PR, but it's something that it may be worth thinking about.

File: kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs
Comment (celinval) [inline kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs:1]: I think there's a lot of boiler plate code here that is duplicated from the cprover version. It may be worth factoring somethings out to `kani_middle` to reduce the burden. Also not a blocker for this PR.

File: kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs
Comment (celinval) [inline kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs:197]: Are we using `llbc` or `aeneas`? We should try to be coherent and only use one of them, don't you think?

File: kani-driver/src/args/mod.rs
Comment (celinval) [inline kani-driver/src/args/mod.rs:1]: Do we want to throw an error if users pass flags that are CBMC specific?

File: kani-driver/src/call_single_file.rs
Comment (celinval) [inline kani-driver/src/call_single_file.rs:107]: Why do we return `None` here but `cprover` on line 59?

File: kani-compiler/src/args.rs
Comment (celinval) [inline kani-compiler/src/args.rs:87]: Please document. Also, what do you think about:

```suggestion
    #[clap(long = "backend", default_value_t = BackendOption::CProver)]
    pub backend: BackendOption,
```

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (celinval) [inline kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:2]: Should this be a modification copyright?

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (celinval) [inline kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:148]: I'm confused. Why can't we use `instance.mangled_name()`?

**Review by celinval — COMMENTED** [review]
A few more comments

File: kani-compiler/Cargo.toml
Comment (zhassan-aws) [reply kani-compiler/Cargo.toml:13]: Good question. I very much prefer using a submodule, and that's how I had it initially, but I couldn't get it to pass CI because of clippy issues in Charon (clippy checks the Charon crate as it is considered part of the Kani workspace). I could not find a way to disable running clippy on a specific crate (the exclude option seems to be broken: https://github.com/rust-lang/rust-clippy/issues/9555), and I eventually gave up.

I'll try to open a PR to address the clippy issues in the library, and once this is done, we can switch to using a git submodule.

File: kani-compiler/src/args.rs
Comment (zhassan-aws) [reply kani-compiler/src/args.rs:90]: This is for testing purposes as part of the `expected` test suite. Since Aeneas is not yet called by the driver, the only option I could think of to test the code is to check that the LLBC is as expected.

File: kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs
Comment (zhassan-aws) [reply kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs:217]: Good point. For Aeneas, I was thinking that the public function mode would be the default.

File: kani-driver/src/args/mod.rs
Comment (zhassan-aws) [reply kani-driver/src/args/mod.rs:1]: Yes. I'll add a TODO.

File: kani-compiler/Cargo.toml
Comment (Nadrieril) [reply kani-compiler/Cargo.toml:13]: Happy to fix the clippy issues and run clippy inour CI (as long as you don't have too pedantic settings on). The main issue is that you're using a somewhat old charon commit so you'd have to upgrade to get these fixes

Comment (zhassan-aws) [reply kani-compiler/Cargo.toml:13]: Thanks @Nadrieril. I'll try to reproduce the clippy issues I ran into and create a PR that fixes them or at least post them somewhere.

Comment (zhassan-aws) [reply kani-compiler/Cargo.toml:13]: I (re-)tried to use a git submodule, and found out that the blocker was not the clippy issues in Charon, but rather that there were compile errors because clippy tries to compile `hax` which is not compatible with our rust toolchain version:
```bash
$ cargo clippy --all  -- -D warnings 
    Checking hax-frontend-exporter v0.1.0-pre.1 (https://github.com/hacspec/hax?branch=main#29d28fcb)
    Checking cprover_bindings v0.55.0 (/home/ubuntu/git/kani/cprover_bindings)
   Compiling kani-compiler v0.55.0 (/home/ubuntu/git/kani/kani-compiler)
    Checking compiletest v0.0.0 (/home/ubuntu/git/kani/tools/compiletest)
    Checking kani_metadata v0.55.0 (/home/ubuntu/git/kani/kani_metadata)
    Checking kani-driver v0.55.0 (/home/ubuntu/git/kani/kani-driver)
error[E0599]: no method named `predicates_defined_on` found for struct `rustc_middle::ty::TyCtxt<'tcx>` in the current scope
   --> /home/ubuntu/.cargo/git/checkouts/hax-310c6deb65d920f9/29d28fc/frontend/exporter/src/state.rs:300:26
    |
300 |             clauses: tcx.predicates_defined_on(did).predicates.sinto(s),
    |                          ^^^^^^^^^^^^^^^^^^^^^
    |
help: there is a method `predicates_of` with a similar name
    |
300 |             clauses: tcx.predicates_of(did).predicates.sinto(s),
    |                          ~~~~~~~~~~~~~

error[E0023]: this pattern has 1 field, but the corresponding tuple variant has 2 fields
    --> /home/ubuntu/.cargo/git/checkouts/hax-310c6deb65d920f9/29d28fc/frontend/exporter/src/types/copied.rs:1290:14
     |
1290 |     Lifetime(Symbol),
     |              ^^^^^^ expected 2 fields, found 1
    --> /home/ubuntu/.rustup/toolchains/nightly-2024-09-12-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_ast/src/token.rs:334:22
     |
     = note: tuple variant has 2 fields
     |
help: use `_` to explicitly ignore each field
     |
1290 |     Lifetime(Symbol, _),
     |                    +++
</snip>
error: could not compile `hax-frontend-exporter` (lib) due to 16 previous errors
```
Apparently, clippy attempts to build the entire Charon package, and not just the library?

Comment (zhassan-aws) [reply kani-compiler/Cargo.toml:13]: OK, I got it to work using:
```
cargo clippy --all --exclude charon --no-deps -- -D warnings
```
I'll switch to using a git submodule.

Comment (zhassan-aws) [reply kani-compiler/Cargo.toml:13]: Opened https://github.com/AeneasVerif/charon/pull/381 to fix the clippy issues I ran into.

Comment (zhassan-aws) [reply kani-compiler/Cargo.toml:13]: Oh, and also ran into errors due to `cargo deny` 😬 

I'll need to update `deny.toml` to work around those.

Comment (zhassan-aws) [reply kani-compiler/Cargo.toml:13]:
```
$ cargo deny --workspace --all-features check advisories
error[unmaintained]: ansi_term is Unmaintained
  ┌─ /home/ubuntu/git/kani/Cargo.lock:5:1
  │
5 │ ansi_term 0.12.1 registry+https://github.com/rust-lang/crates.io-index
  │ ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ unmaintained advisory detected
  │
  ├ ID: RUSTSEC-2021-0139
  ├ Advisory: https://rustsec.org/advisories/RUSTSEC-2021-0139
  ├ The maintainer has advised that this crate is deprecated and will not receive any maintenance.
    
    The crate does not seem to have much dependencies and may or may not be ok to use as-is.
    
    Last release seems to have been three years ago.
    
    ## Possible Alternative(s)
    
     The below list has not been vetted in any way and may or may not contain alternatives;
    
     - [ansiterm](https://crates.io/crates/ansiterm)
     - [anstyle](https://github.com/epage/anstyle)
     - [console](https://crates.io/crates/console)
     - [nu-ansi-term](https://crates.io/crates/nu-ansi-term)
     - [owo-colors](https://crates.io/crates/owo-colors)
     - [stylish](https://crates.io/crates/stylish)
     - [yansi](https://crates.io/crates/yansi)
    
    ## Dependency Specific Migration(s)
    
     - [structopt, clap2](https://github.com/clap-rs/clap/discussions/4172)
  ├ Announcement: https://github.com/ogham/rust-ansi-term/issues/72
  ├ Solution: No safe upgrade is available!
  ├ ansi_term v0.12.1
    └── assert_tokens_eq v0.1.0
        └── (dev) macros v0.1.0
            └── charon v0.1.36
                └── kani-compiler v0.55.0

error[unmaintained]: difference is unmaintained
   ┌─ /home/ubuntu/git/kani/Cargo.lock:56:1
   │
56 │ difference 2.0.0 registry+https://github.com/rust-lang/crates.io-index
   │ ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ unmaintained advisory detected
   │
   ├ ID: RUSTSEC-2020-0095
   ├ Advisory: https://rustsec.org/advisories/RUSTSEC-2020-0095
   ├ The author of the `difference` crate is unresponsive.
     
     Maintained alternatives:
     
     - [`dissimilar`](https://crates.io/crates/dissimilar)
     
     - [`similar`](https://crates.io/crates/similar)
     
     - [`treediff`](https://crates.io/crates/treediff)
     
     - [`diffus`](https://crates.io/crates/diffus)
   ├ Announcement: https://github.com/johannhof/difference.rs/issues/45
   ├ Solution: No safe upgrade is available!
   ├ difference v2.0.0
     └── assert_tokens_eq v0.1.0
         └── (dev) macros v0.1.0
             └── charon v0.1.36
                 └── kani-compiler v0.55.0

advisories FAILED
```

Comment (zhassan-aws) [reply kani-compiler/Cargo.toml:13]: One more: `kani-regression.sh` fails because of unformatted Charon code. I'll need to add an entry in `rustfmt.toml` to ignore `charon`.

Comment (zhassan-aws) [reply kani-compiler/Cargo.toml]: Sooooooo, to summarize, adding Charon as a git submodule has 3 implications:
- It gets checked by clippy
- It gets checked by cargo fmt
- It gets checked by cargo deny
I worked around all 3 issues, but we should fix the issues upstream and remove the workarounds.

Comment (Nadrieril) [reply kani-compiler/Cargo.toml]: The cargo deny issues look easy to fix. Clippy I can add to our CI. Formatting I thought we already had this in our CI, what was the diff?

Comment (zhassan-aws) [reply kani-compiler/Cargo.toml]: Here's the diff:

https://gist.github.com/zhassan-aws/d72bacfcb7bc93a94fd978d9feb4b71e

Comment (celinval) [reply kani-compiler/Cargo.toml]: Having `cargo deny` failing is not a bad thing, is it? I'm actually surprised it wasn't failing before, doesn't it check the entire dependency tree? You should be able to go around `cargo fmt` and `cargo deny`, since this should not be a crate in the workspace.

File: kani-compiler/src/args.rs
Comment (celinval) [reply kani-compiler/src/args.rs:90]: I think this has to at least be marked as a hidden option, since it is not a behavior we want to keep. This should also be documented as a temporary option, and that it requires `-Z aenas`.

For these things, we usually use the `script-based-pre` tests, but depending on the number of tests you want to add, it can become very painful to maintain. So I'll leave it up to you.

File: kani-compiler/Cargo.toml
Comment (zhassan-aws) [reply kani-compiler/Cargo.toml]: @Nadrieril it looks like the formatting is due to the configuration we're using in https://github.com/model-checking/kani/blob/main/rustfmt.toml

@celinval do you know why we're not using the default formatting configuration?

File: kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs
Comment (celinval) [reply kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs:191]: I wouldn't use CBMC as a model of consistency. :laughing:

I think this is a usability question, and being consistent and using a meaningful name might be the way to go here. I don't think any of our users really know what `goto` is, and using it in a user visible manner is probably more confusing than anything else. Try googling "goto" and you'll see what I'm saying. LOL

My main question to you is, what feature are the users enabling, `aeneas`, `llbc` or `lean`? Which ones are they going to be interacting with?

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (zhassan-aws) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:148]: My understanding is that this tries to preserve the original names as much as possible so that it's easy for the user to relate the generated code to the original Rust.

File: kani-driver/src/args/mod.rs
Comment (zhassan-aws) [reply kani-driver/src/args/mod.rs:1]: Done.

File: kani-driver/src/call_single_file.rs
Comment (zhassan-aws) [reply kani-driver/src/call_single_file.rs:105]: Good catch. I updated the one on line 59.

File: kani-compiler/src/args.rs
Comment (zhassan-aws) [reply kani-compiler/src/args.rs]: Makes sense. Updated.

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (zhassan-aws) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs]: Updated.

File: kani-compiler/Cargo.toml
Comment (zhassan-aws) [reply kani-compiler/Cargo.toml]: > Having `cargo deny` failing is not a bad thing, is it? I'm actually surprised it wasn't failing before, doesn't it check the entire dependency tree? You should be able to go around `cargo fmt` and `cargo deny`, since this should not be a crate in the workspace.

Good question. It was previously added in the `allow-git` section of the `deny.toml`:
```
allow-git = [
    "https://github.com/AeneasVerif/charon",
]
```
Perhaps this makes `cargo deny` not look at its dependencies?

File: kani-compiler/src/args.rs
Comment (zhassan-aws) [reply kani-compiler/src/args.rs:92]: > I think this has to at least be marked as a hidden option, since it is not a behavior we want to keep.

I recall trying to hide the option, but I couldn't find a way to do that. I'll check if that's possible.

> and that it requires -Z aenas.

I added that to the option documentation. The option validation would already error out if the option is specified without `-Zaeneas`.

> For these things, we usually use the script-based-pre tests, but depending on the number of tests you want to add, it can become very painful to maintain. So I'll leave it up to you.

Yeah, I considered this option, but it was indeed going to be painful.

File: kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs
Comment (Nadrieril) [reply kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs:191]: I'd say it depends on whether you want to use other tools than Aeneas on the output llbc. We have Eurydice that compiles llbc to C and we're developing simple static analyses at the moment.

Comment (zhassan-aws) [reply kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs:191]: @celinval My thinking was that the kani-compiler is an internal binary that is not meant to be used directly. What the kani-compiler is concerned with is to take MIR and translate it to GOTO or LLBC. The kani-driver is the  binary users interact with, and thus the options at that level should mention Lean or Aeneas.

What do you think?

File: kani-compiler/Cargo.toml
Comment (zhassan-aws) [reply kani-compiler/Cargo.toml]: > @celinval do you know why we're not using the default formatting configuration?

Answering my own question: it seems we inherited the non-default configuration from rustc.

File: kani-compiler/src/args.rs
Comment (zhassan-aws) [reply kani-compiler/src/args.rs:92]: The option is now hidden.

File: kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs
Comment (celinval) [reply kani-compiler/src/codegen_aeneas_llbc/compiler_interface.rs:191]: For internal names in the compiler, I think `llbc` or `aeneas` would make sense. For user visible things, I think it will depend on what will be visible for the user. If users are not expected to interact with aeneas directly, I would prefer using Lean.

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (celinval) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:150]: I think it would be great to get more clarity on why we need this extra complexity. I'm also concerned that this could impact the analysis soundness (in case of 2+ names being resolved to the same name). It's also not clear to me how this is handling generic parameters.

Comment (celinval) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:150]: Another potential issue is with different crate versions.. I wonder if they will have the same name for multiple items.

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (celinval) [inline kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:449]: I don't think you should be ignoring `args` here. You need to instantiate the type even when handling generic code.

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (celinval) [inline kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:381]: From our offline chat. I think what you want is the following, you only want to generate code for generic code, but you want to use the reachability code to collect everything that is reachable from the user input. In that case, I recommend you do the following.

1. After the collector runs, build a set of `CrateDef` from all the collected instances. You can convert an instance to a `CrateItem` using the `TryFrom` implementation [here](https://github.com/rust-lang/rust/blob/baaf3e65abab5a281715230bcb1785610629ae7d/compiler/stable_mir/src/mir/mono.rs#L214-L226).
2. Most places where you use instance, you can replace by the crate item. I think the only place that you will need to change is where you the the `fn_abi()`. In that case, you need to use the function signature instead.

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (celinval) [inline kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:133]: Why do you need internal id here?

**Review by celinval — COMMENTED** [review]
I think it would be great if we could eventually simplify the naming schema, and remove the name generation code. This seems fairly fragile, hard to maintain and I worry that this will not work (or even be unsound) if 2+ crates have the same name. That said, I think it is out of scope of this PR since you are only trying to match Aeneas name schema.

The handling of monomorphic vs generic code also need some improvement. To unblock this PR, can we add an error if we encounter generic code? I think none of your tests depend on generic code.

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (zhassan-aws) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:133]: Charon's data structures are in terms of the internal `DefId`: https://github.com/AeneasVerif/charon/blob/4caeeb35331b7e6ee59565a3eb698a200c6e90e9/charon/src/ast/krate.rs#L110

This was removed recently (in https://github.com/AeneasVerif/charon/commit/8a80793e440b256611baf747483a1b24613a043e), so once we upgrade to that commit, we should no longer need it.

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (zhassan-aws) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs]: I added an `unimplemented` for non-empty `args`. Would that work?

**Review by celinval — APPROVED** [review]
Sweet! I this is a great first step. Thanks @zhassan-aws

File: kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs
Comment (Nadrieril) [reply kani-compiler/src/codegen_aeneas_llbc/mir_to_ullbc/mod.rs:150]: These names are used in two ways: as @zhassan-aws said we use them when displaying info to the user so they can recognize which item we're talking about, and we also sometimes pattern-match on them to recognize built-ins. As of latest charon we use lang/diagnostic items wherever possible, but we still rely on names to recognize e.g. `core::panicking::panic_explicit`. That's not crucial.

Generic parameters are only relevant for inherent impl blocks. We include them in the name:
```rust
impl<T: Clone> Foo<T> {
    fn method(&self) {}
}
```
This method has name `crate::module::{impl<T: Clone> Foo<T>}::method`.

Well spotted on crate versions, I opened an [issue](https://github.com/AeneasVerif/charon/issues/412) for that.

**Pull request merged**
